File: go/wrench.go
// Package wrench reads, writes and validates the form of the ecosystem's
// structured files.
//
// File handling is two calls. Validation sits in the signature, so nothing
// reads or writes without naming what the file must conform to. The codec is
// the format and the reader or writer is the IO, declared separately, which
// puts the IO boundary wholly outside the call.
//
// The contract names these calls load_formatted_file and save_formatted_file.
// A pack spells them the way its own language spells things, so the Go pack
// exports LoadFormattedFile and SaveFormattedFile. What the packs share is
// behaviour, not spelling.
package wrench

// A usageError is a call made wrongly, before any file was touched. It is a
// seventh kind beside the six steps because it is not one of them: nothing was
// read, parsed or validated.
//
// These stay sentinel values so errors.Is keeps working, and gain Step so
// errors.As reaches them through the Error interface. Before this they were
// plain errors.New values, so a consumer catching "any wrench failure" missed
// exactly the failures that mean it called wrench wrong.
type usageError struct{ message string }

func (e *usageError) Error() string { return e.message }

func (e *usageError) Step() string { return StepUsage }

// ErrNoSchema is returned when a call is handed no schema. The signature
// compels one; nil is the one way round that, and it is refused here so
// FR-2.3's guarantee holds in a language where any interface can be nil.
var ErrNoSchema error = &usageError{"wrench: no schema given"}

// ErrNoCodec is returned when a call is handed no codec.
var ErrNoCodec error = &usageError{"wrench: no codec given"}

// ErrNoReader is returned when a load is handed no reader.
var ErrNoReader error = &usageError{"wrench: no reader given"}

// ErrNoWriter is returned when a save is handed no writer.
var ErrNoWriter error = &usageError{"wrench: no writer given"}

// A Codec turns bytes into a structure and a structure into canonical bytes.
// It is the format, and knows nothing about where the bytes came from.
type Codec interface {
	// Decode turns a file's bytes into maps, lists and scalars.
	Decode(data []byte) (any, error)
	// Encode turns a structure into bytes in the codec's canonical form.
	Encode(value any) ([]byte, error)
}

// A Reader hands back the whole contents of a named file. It is the IO, and
// knows nothing about the format. A Reader is handed the path rather than
// bytes, so substituting one in a test exercises the validation paths against
// no filesystem at all.
type Reader interface {
	Read(path string) ([]byte, error)
}

// A Writer puts the whole contents of a named file in place.
type Writer interface {
	Write(path string, data []byte) error
}

// A Schema validates a decoded structure. It applies to the maps and lists a
// codec produced rather than to the text, so it is indifferent to how the file
// was serialised on the way in.
type Schema interface {
	Validate(value any) error
}

// LoadFormattedFile reads path through reader, decodes it with codec, and
// validates the result against schema.
//
// A failure says which step failed. ReadError, ParseError and ValidationError
// are distinct because they have distinct causes and distinct fixes.
func LoadFormattedFile(path string, schema Schema, codec Codec, reader Reader) (any, error) {
	if err := requireLoadParts(schema, codec, reader); err != nil {
		return nil, err
	}

	data, err := reader.Read(path)
	if err != nil {
		// Already wrench's, from a codec or a schema that has no path.
		// Fill it in rather than wrap a second time.
		if filled, ours := atPath(err, path); ours {
			return nil, filled
		}
		return nil, &ReadError{Path: path, Err: err}
	}

	value, err := codec.Decode(data)
	if err != nil {
		// Already wrench's, from a codec or a schema that has no path.
		// Fill it in rather than wrap a second time.
		if filled, ours := atPath(err, path); ours {
			return nil, filled
		}
		return nil, &ParseError{Path: path, Err: err}
	}

	if err := schema.Validate(value); err != nil {
		// Already wrench's, from a codec or a schema that has no path.
		// Fill it in rather than wrap a second time.
		if filled, ours := atPath(err, path); ours {
			return nil, filled
		}
		return nil, &ValidationError{Path: path, Err: err}
	}

	return value, nil
}

// SaveFormattedFile validates value against schema, encodes it with codec in
// canonical form, and writes it to path through writer.
//
// Validation runs before the write, so a caller cannot put down a structure
// wrench would refuse to read back.
func SaveFormattedFile(value any, path string, schema Schema, codec Codec, writer Writer) error {
	if err := requireSaveParts(schema, codec, writer); err != nil {
		return err
	}

	if err := schema.Validate(value); err != nil {
		// Already wrench's, from a codec or a schema that has no path.
		// Fill it in rather than wrap a second time.
		if filled, ours := atPath(err, path); ours {
			return filled
		}
		return &ValidationError{Path: path, Err: err}
	}

	data, err := codec.Encode(value)
	if err != nil {
		// Already wrench's, from a codec or a schema that has no path.
		// Fill it in rather than wrap a second time.
		if filled, ours := atPath(err, path); ours {
			return filled
		}
		return &EncodeError{Path: path, Err: err}
	}

	if err := writer.Write(path, data); err != nil {
		// Already wrench's, from a codec or a schema that has no path.
		// Fill it in rather than wrap a second time.
		if filled, ours := atPath(err, path); ours {
			return filled
		}
		return &WriteError{Path: path, Err: err}
	}

	return nil
}

func requireLoadParts(schema Schema, codec Codec, reader Reader) error {
	if schema == nil {
		return ErrNoSchema
	}
	if codec == nil {
		return ErrNoCodec
	}
	if reader == nil {
		return ErrNoReader
	}
	return nil
}

func requireSaveParts(schema Schema, codec Codec, writer Writer) error {
	if schema == nil {
		return ErrNoSchema
	}
	if codec == nil {
		return ErrNoCodec
	}
	if writer == nil {
		return ErrNoWriter
	}
	return nil
}

File: ruby/lib/wrench.rb
# frozen_string_literal: true

require_relative "wrench/errors"
require_relative "wrench/float_text"
require_relative "wrench/schema"
require_relative "wrench/seams"
require_relative "wrench/json_codec"
require_relative "wrench/yaml_codec"

# wrench reads, writes and validates the form of the ecosystem's structured
# files. This is the Ruby pack.
#
# THE TWO CALLS ARE THE WHOLE OF FILE HANDLING, FR-2.1:
#
#     load_formatted_file(path, schema, codec, reader) -> value
#     save_formatted_file(value, path, schema, codec, writer)
#
# Load reads, decodes, then validates. Save validates, encodes, then writes.
# Validating on the way out is not symmetry for its own sake: it stops a caller
# writing a structure wrench would refuse to read back, FR-2.4, so a file
# produced by a save always survives a load.
#
# The schema argument cannot be omitted, FR-2.2. It can be wrong, and no part of
# this library detects that, FR-2.3.
#
# TWO CODECS SHIP, YAML AND JSON. TOML was the third and is retired, FR-2.7: no
# maintained library in any of the four languages could emit its canonical form,
# and the hand-written emitters written instead wrote documents they could not
# read back.
module Wrench
  # The codecs, named rather than inferred from a filename. Choosing a parser by
  # suffix would make behaviour depend on what a file is called, and renaming a
  # file would silently change how it is read.
  YAML = YAML_CODEC
  JSON = JSON_CODEC

  module_function

  # Read `path` through `reader`, decode it with `codec`, validate the result.
  #
  # Each step's failure is its own kind, so a caller learns whether the file was
  # unreadable, unparseable or simply wrong. A codec is handed bytes and a
  # schema a structure, so neither knows which file it is working on: both raise
  # with no path and this fills it in.
  def load_formatted_file(path, schema, codec, reader)
    require_seams(schema: schema, codec: codec, io: reader)
    bytes = reader.read(path)
    value = with_path(path) { codec.decode(bytes) }
    with_path(path) { schema.validate(value) }
    value
  end

  # Validate `value`, encode it with `codec`, and put it in place through
  # `writer`.
  def save_formatted_file(value, path, schema, codec, writer)
    require_seams(schema: schema, codec: codec, io: writer)
    with_path(path) { schema.validate(value) }
    bytes = with_path(path) { codec.encode(value) }
    writer.write(path, bytes)
  end

  def load_yaml_file(path, schema, reader) = load_formatted_file(path, schema, YAML, reader)
  def save_yaml_file(value, path, schema, writer) = save_formatted_file(value, path, schema, YAML, writer)
  def load_json_file(path, schema, reader) = load_formatted_file(path, schema, JSON, reader)
  def save_json_file(value, path, schema, writer) = save_formatted_file(value, path, schema, JSON, writer)

  # A missing seam is `usage`, the kind that names no step that ran, because the
  # call was made wrongly before any file was touched.
  def require_seams(schema:, codec:, io:)
    raise UsageError, "no schema was given" if schema.nil?
    raise UsageError, "no codec was given" if codec.nil?
    raise UsageError, "no reader or writer was given" if io.nil?
  end

  # Attach the file to an error raised by something that does not know it.
  def with_path(path)
    yield
  rescue Error => e
    raise e.class.new(e.message, cause: e.cause_error, path: path)
  end
end
